### DISCUSSION TITLE
Can I have `turbo build` include an individual "apps" dependencies in the `dist`?

### DISCUSSION THREAD

**Discussion by kevin-mitchell** [vercel/turborepo#9019] [Help]
### Summary

I'm using `turborepo` to build individual Lambda functions, but would like the build output for each app to (optionally) include all of the dependencies from each individual package.json file in node_modules. Looking for some guidance. I'm fairly certain based on my googling so far that "this is the job of a bundler" is the answer, but I'm hoping for a bit more guidance here if possible!

### Additional information

I have a turborepo setup where each `app` is a different AWS Lambda function (the fact this is Lambda doesn't really matter, but just for context).

I am writing everything in TypeScript, and each "app" is very simple generally (one, maybe two TS files). For the most part I have no external dependencies because I'm using the AWS SDK which is available in the Lambda runtime, so I just have it set as a dev dependency for now (again, this doesn't matter but just for context).

I now have a situation where I need to use an external dependency, and my package.json file looks like this:

```json
{
  "name": "lambda-api-server",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "compile": "tsc",
    "build": "tsc",
    "dev": "npm run compile && node ./dist/index.js",
    "test": "echo 'ok'"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "type": "module",
  "devDependencies": {
    "@aws-sdk/client-sns": "^3.629.0",
    "@repo/eslint-config": "*",
    "@repo/typescript-config": "*",
    "@types/node": "^20.12.7",
    "aws-cdk": "^2.141.0",
    "aws-cdk-lib": "^2.141.0",
    "typescript": "^5.4.5"
  },
  "dependencies": {
    "graphql": "^16.8.1",
    "ksuid": "^3.0.0"
  }
}
```

`ksuid` and `graphql` (and their dependency trees) I need to be included with my deployed function and ideally that would mean when I run `turbo build` somehow, "magically", turbo does whatever it needs to do to move the dependencies from the shared root node_modules into the `dist`.

It's outside of the scope of turbo again but for context I'm essentially running `turbo build` followed by `cdk deploy`, and each lambda function is being deployed from the `dist` directory. So again ideally after `turbo build` all required dependencies would be included in the `dist`.

**I am fairly certain the fact that I'm asking this question means I'm "doing it wrong", but am still hoping for some advice as to how I should be handling this setup** - to be entirely honest I'm not even sure using turborepo is the right choice for me here but I've enjoyed working with it in the past with a NextJS project and other than this bundling issue it's been working great for me to manage a lot of individual small apps.

Thank you very much!
```


### Example

_No response_

**Comment by hsellik** [comment]
I assume that you have package.json file for each app separately and workspaces are set up correctly in the root package.json. Isn't `turbo build` already including all the required dependencies in the dist folder by default? 

It should be more like yarn/npm/pnpm workspaces magic rather than turbo magic.

**Reply by kevin-mitchell** [reply]
Thanks for the reply!

So I decided to try to start from a simpler project and see if I was missing something and sure enough, until I introduced something like `rollup` or `esbuild` or another bundler I was not ending up with my dependencies included in the project.

So I went back and setup esbuild in each of my "apps" and that resolved my issue. I don't end up with `node_modules` but that's probably fine / for the best (?), but my dependencies are all bundled with my final build artifact (i.e. index.ts -> index.js which has all of the dependencies included).

**Comment by kevin-mitchell** [comment]
In the end I realized even a simple project without turborepo, just a flat `npm init` project with typescript, I don't magically get my dependencies included for me, I have to (obvious I realize) include some other build step, i.e. a bundler such as `esbuild` or `rollup` (I ended up using `esbuild`).

In my simple case it was enough to

1.  `npm install esbuild esbuild-plugin-tsc --save-dev`
1. and add a file `esbuild.config.mjs` with this config:

```javascript
import * as esbuild from 'esbuild'

await esbuild.build({
  entryPoints: ['src/index.ts'],
  bundle: true,
  outfile: 'dist/index.js',
  platform: 'node'
})
```

3. For my build script run `node esbuild.config.mjs`

Again this is really outside of the concern of turborepo, but hopefully somebody in a similar confused state might find this.
